fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+8'b10101010
+```
+
+For Verilog style constants the '?' value indicates a match all or
+don't care element. The width of '?' is relative to the width of 
+native elements within the constant's base.
+
+```
+ 8'b1010101?      bit 0 is a don't care
+12'h45?           bits [3:0] are a don't care
+```
+
+----------------------------------------------------------------
+## Operators
+
+The constraints clause is the target use case for defining operator
+support in FSL.
+
+The list of operators and their support in FSL is listed below. FSL adds
+the range operator to the standard C style operators. The syntax is 
+similar to Verilog.
+
+These operators are unnecessary in FSL: Divide, modulus, size_of,
+pointer operations, type conversions cast/implicit/explicit.
+
+The comma operator support is limited to concatentaion expressions, operation
+chaining is not necessary in FSL.
+
+Within the supported operators FSL uses conventional operator precedence.
+
+
+### Arithmetic Operators
+
+```
+    + (Addition)
+    - (Subtraction)
+    * (Multiplication)
+```
+
+### Range Operators
+
+```
+    [M:N] (range select)
+    [M]   (index)
+```
+
+### Relational Operators
+
+```
+    ==  (Equal to)
+    !=  (Not equal to)
+    >   (Greater than)
+    <   (Less than)
+    >=  (Greater than or equal to)
+    <=  (Less than or equal to)
+```
+
+### Logical Operators
+
+```
+    &&  (Logical AND)
+    ||  (Logical OR)
+    !   (Logical NOT)
+```
+
+
+### Assignment Operator
+
+```
+    =    (Simple assignment)
+```
+
+
+### Concatenation Operator
+
+Concatenation expressions use a style similar to verilog, {a,b,c};
+
+```
+  xyz = { 3'b00, 4'h3,8'd10 }
+  abc = { a, b, c }
+  
+```
+
+## Comments
+
+Comments use '# ' (hash/space) for single line comments or '###' as 
+the delimiter for comment blocks. Comments can be embedded in other comments.
+
+```
+#  ### commenting out
+#  my multi line
+#  comment by
+#  prefixing the '#'
+#  end of ML comment ###
+```
+
+Mixing block and line comments is supported, the first comment token
+takes precedence. This practice is not recommended.
+
+```
+#  ### commenting out
+#  my multi line
+   this line will be processed
+#  prefixing the '#'
+#  end of ML comment ###
+```
+
+In this case the block comment start and stop markers are not processed 
+because they are delimited by line comments.  This is supported but 'this
+line will be processed' will be tokenized and tested for legal grammar. 
+
+## Preprocessor Support
+
+The FSL parser uses a pre-processor with features similar to that found
+in the C pre-processor.  Within the constraints of the FSL language most 
+pre-processor operations found in C will be available in FSL with slight
+changes for FSL syntax choices.
+ 
+Note that the typical # indicator, as in #include, is replaced with
+'`' as in `include.
+
+### Conditional Compilation
+
+FSL supports a common set of conditional compilation directives, but 
+with the typical # replaced by back-tick.
+
+```
+`ifdef X
```

Comment:
This is just preference when implementing the parser/grammar, 

Use of '#' for preprocessor directives dates to K&R supposedly selected because it was not prevalent in any languages at the time. but now #'s are found in some assembly languages. like ARM asm.

At the moment '#' is the current comment character. But this may change.

I chose the verilog syntax because I'm used to it and it's well known and back ticks are not at all common in the assembly languages I have been exposed too.

But again, boils down to my preference.